### ISSUE TITLE
Fall back to Gaussian rules in Lobatto classes for certain element types

### ISSUE THREAD

**Issue by hchen139** [libMesh/libmesh#2545]
First, the rational behind not falling back to Gaussian rules for Lobatto classes is completely understandable. For certain cases, it may be desirable to have integration points exactly at the nodes (such as peridynamics discussed in the following content).

Let me talk about my situation and why I would like to have the Lobatto classes fall back to Gaussian rules (actually any other rules work for the element types in my application would be good) for certain element types. Recently, peridynamics has become a module in the MOOSE framework. Peridynamics theory is a nonlocal integral reformulation of the classical continuum mechanics theory, which has strong advantages in materials failure modeling. Different from the finite element implementation of a continuum mechanics problem, peridynamics discretize the domain of interest into material points with finite volume, and interaction between material points is formulated as a bond-like interaction. In peridynamics, all quantities, such as displacement, strains, and stresses, are defined at a material point. The peridynamics module in the MOOSE framework takes the advantage of two-node Edge element and Lobatto rules to store material properties at the two material points, i.e., the quadrature points. To this point, there were no problems with the 1D Lobatto rules for 2-node Edge in applications using peridynamics.

For contact-like problems, peridynamics relies on the 3-node triangular element (for 2D problems) and 4-node tetrahedral element (for 3D problems) and the available contact algorithms in MOOSE. These higher dimensional elements are constructed based on material points where contact will occur between two dissimilar bodies. These are "dummy" elements only used to obtain the contact force at those material points. The actual underlying physics are still modeled using 2-node Edge elements. For problems like this, therefore, there will be mixed element types in the simulation. Recall the Lobatto rules are needed for peridynamics and Lobatto rules are not implemented or not available for 3-node triangular element and 4-node tetrahedral element, it becomes impossible to run these peridynamic simulations.

For current modeling and simulation, I locally made the default to fall back to Gaussian rules for those elements. This solves all my aforementioned issues. However, for users to use certain advanced capabilities offered in the peridynamics module, it is necessary to have either fall back to Gaussian rules for 3-node triangular element and 4-node tetrahedral element, or implement Lobatto rules for those element types. While the first solution is easy to make, the second solution may be impossible. For the second solution, I search online and cannot find any Lobatto rules for 4-node tetrahedral element, although there are papers talked about Lobatto rules for 3-node triangular element, such as https://arxiv.org/pdf/1007.4848.pdf.

It is preferable to fall back to Gaussian rules for certain element types. Any feedback or suggestion will be much appreciated.

**Comment by jwpeterson** [comment]
It sounds like what you are really after is nodal quadrature? If so, the `QNodal` class may be more appropriate than `QGaussLobatto`, and I believe it is implemented for all the element types which you mentioned. The drawback, of course is that nodal quadratures are less accurate than rules with interior points.

> Lobatto rules are not implemented or not available for 3-node triangular element and 4-node tetrahedral element, it becomes impossible to run these peridynamic simulations.

I think it would be fine if we change the `default` case to fall back on `QGauss` for the unsupported element types. I still agree in principle with the comment that is in the code 
```
// We *could* fall back to a Gauss type rule for other types
// elements, but the assumption here is that the user has asked
// for a Gauss-Lobatto rule, i.e. a rule with integration points
// on the element boundary, for a reason.
```
but in practice the fallback (maybe with a one time only `libmesh_warning`) is probably the right thing to do.

**Comment by hchen139** [comment]
John, many thanks for your quick reply. The only concern to use nodal quadrature is when running simulation with both peridynamics and finite element are used, i.e., peridynamics for part of the domain and finite element for the rest. In cases like this, yes we don't want to sacrifice the accuracy in the finite element domain. In peridynamics, we don't really do accurate numerical integration but treat as a finite sum. Therefore, nodal or elemental integration won't matter for peridynamics.

The solution to fallback to QGauss with one time libmesh warning would be a good solution. Would you mind to make that change, since I don't really have a libmesh repo on my machine?

**Comment by jwpeterson** [comment]
Ah, yes that makes sense for MOOSE since everything should use the same Quadrature rule.

> Would you mind to make that change, since I don't really have a libmesh repo on my machine?
Yep, should just be a few lines... I'll tag you after I create it.

**Comment by jwpeterson** [comment]
Fixed in #2547

**Comment by hchen139** [comment]
@jwpeterson John, thanks for your help to get this done so fast.
